core/src/main/java/com/scalar/db/transaction/consensuscommit/RollbackMutationComposer.java

```diff
@@ -84,28 +85,24 @@ public void add(Operation base, TransactionResult result) {
     }
   }
 
-  private Put composePut(Operation base, TransactionResult result) {
+  private Put composePut(Operation base, TransactionResult result) throws ExecutionException {
     assert result.getState().equals(TransactionState.PREPARED)
         || result.getState().equals(TransactionState.DELETED);
 
+    TransactionalTableMetadata metadata = tableMetadataManager.getTransactionalTableMetadata(base);
+    LinkedHashSet<String> beforeImageColumnNames = metadata.getBeforeImageColumnNames();
+
     Map<String, Value<?>> map = new HashMap<>();
     result
         .getValues()
         .forEach(
             (k, v) -> {
-              if (k.startsWith(Attribute.BEFORE_PREFIX)) {
+              if (beforeImageColumnNames.contains(k)) {
                 String key = k.substring(Attribute.BEFORE_PREFIX.length());
                 map.put(key, v.copyWith(key));
               }
             });
 
-    // remove keys
-    Stream.of(
```

Comment:
Thank you for taking a look at this!

The change makes it not needed because `beforeImageColumnNames` doesn't have primary key (partition key and clustering key) columns. So `map` should not have primary key columns.

Reply:
Thank you!